File: client/src/app/utillity/Event.ts
export default class Event {

    private static listeners: Set<EventHandler>;

    static listen(name, callback) {
        if (Event.listeners === undefined) {
            Event.listeners = new Set<EventHandler>();
        }
        
        Event.listeners.add({ name, callback });
    }


    static emit(name, ...data) {
        if (Event.listeners === undefined) {
            Event.listeners = new Set<EventHandler>();
        }

        Event.listeners.forEach(listener => {
            if (listener.name === name) {
                listener.callback(data);
            }
        })
    }
}

class EventHandler {
    name: string
    callback: Function;
}

File: server/src/main/java/shared/event/Event.java
package shared.event;

import shared.event.handler.EventHandler;
import shared.util.Log;

import java.util.*;


public class Event {

    private static Map<String, List<EventHandler>> listeners = new HashMap<>();

    public static void listen(String name, EventHandler eventHandler) {

        Log.debug("Event", "Started listening for '" + name + "' with eventHandler: " + eventHandler);

        List<EventHandler> eventHandlers = listeners.get(name);

        if (eventHandlers == null) {
            listeners.put(name, Collections.singletonList(eventHandler));
        } else {
            eventHandlers.add(eventHandler);
        }
    }


    public static void emit(String name, Object... data) {

        Log.debug("Event", "Emitting '" + name + "' with data: " + Arrays.deepToString(data));

        listeners.forEach((listenerName, handlers) -> {
            if (listenerName.equals(name)) {
                handlers.forEach(handler -> handler.handle(name, data));
            }
        });
    }
}


